### ISSUE TITLE
Module version SHA pinning

### ISSUE THREAD

**Issue by nitrocode** [opentofu/opentofu#1942]
### OpenTofu Version

```shell
Latest
```

### Use Cases

Some organizations have a distrust of upstream tampering of versions. One way to give peace of mind is by pinning versions to upstream shas.

With a more explicit decision, tools for dependency management (such as renovate and dependabot) can start supporting SHA pinning for terraform modules for security hardening.

Currently, to use `version` meta argument in module definitions, only a published tag can be specified (afaik).

```hcl
module "ecr" {
  source  = "cloudposse/ecr/aws"
  version = "0.42.0"

  name = "ecr"
}
```

### Attempted Solutions

An alternative would be

```hcl
module "ecr" {
  source = "git::https://github.com/cloudposse/terraform-aws-ecr.git?ref=2896ea4fd21a41ebe86d6bed7f68735a51e83831
" # v0.42.0
}
```

### Proposal

Id prefer if we can do something like github actions

```yaml
    steps:
      - uses: actions/checkout@db41740e12847bb616a339b75eb9414e711417df # v1.2.3
```

Or docker

```docker
FROM python:3.11.3-alpine@sha256:06a3f7b12a56ed051612416d3083814e145afc7fc3fc52a67aeffa348a20383c
```

---

Option 1) Use tag and ref (similar to gha and docker)

```hcl
module "ecr" {
  source  = "cloudposse/ecr/aws"
  version = "0.42.0@2896ea4fd21a41ebe86d6bed7f68735a51e83831"
}
```

Option 2) `version` is an alias for git's `ref` with a comment for the tag

```hcl
module "ecr" {
  source  = "cloudposse/ecr/aws"
  version = "2896ea4fd21a41ebe86d6bed7f68735a51e83831
" # v0.42.0
}
```

Option 3) use the alternative via dropping version for source git and ref then manually add a comment for the version

### References

- https://docs.github.com/en/actions/security-for-github-actions/security-guides/security-hardening-for-github-actions
- https://developer.hashicorp.com/terraform/language/modules/sources#generic-git-repository
- https://docs.renovatebot.com/docker/#digest-pinning

**Comment by cam72cam** [comment]
As far as the registry goes, we should be storing and checking the tag against the commit hash and returning that.  Tags in the registry should be idempotent.

I could have swore we had an issue in the opentofu/registry repo, but I'm not finding it.

Would that change to the registry satisfy this issue @nitrocode?

**Comment by nitrocode** [comment]
Yes agreed. The concern is that the tag is a mutable property and this can be overwritten so pinning to a SHA is far more difficult to overwrite.

**Comment by stevehipwell** [comment]
Isn't this use case covered by #586?

**Comment by nitrocode** [comment]
That's a good call.

Kind of. A lot of us do not commit the lockfile. I'd prefer the pin to be in the hcl itself if feasible.

**Comment by stevehipwell** [comment]
@nitrocode why wouldn't you commit the lock file? IaC is meant to be deterministic but without a checked in lock file it can't be (unless you're using explicit provider versions at which point why not use the lock file).

**Comment by nitrocode** [comment]
I previously had issues in our org when it came to lock files between osx intel, osx arm, and linux x86.

E.g. Id commit the file using osx arm then the x86 linux pod (where terraform runs) throws errors because it doesn't understand the osx arm architecture pinned.

I suppose we can each run this but then we'd have to train devs to run this every time they create a new terraform directory

```
terraform providers lock \
    -platform=darwin_amd64 \
    -platform=linux_amd64 \
    -platform=darwin_arm64 \
    -platform=linux_arm64
```

**Comment by stevehipwell** [comment]
@nitrocode that is the current (complex) spec for correctly generating a lock file for multiple architectures.

Wouldn't opening an issue here to support defining lock file architectures in a workspace or globally be a better approach than attempting to work around this be not using lock files?

**Comment by nitrocode** [comment]
Fair point. If locking files is easier for developers without having to know what architectures need locking for it to run in CICD and locally and locking modules can exist in the lockfile, then we do not need to pin SHAs within the HCL code itself.

One could also argue that the lockfile for providers can be replaced by using a similar convention above.

```terraform
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "360d526b37cd514d9544fdc199457948fc4f7ef7" # v5.64.0
    }
  }
}
```

https://github.com/hashicorp/terraform-provider-aws/releases/tag/v5.64.0

**Comment by stevehipwell** [comment]
@nitrocode using a digest as a version isn't the same as using a lock file containing a digest. A version digest tracks the source code but the lock file carries a digest for the built binaries. Your example above wouldn't stop a malicious actor from modifying the provider payload attached to the immutable version digest.

FYI this is still relevant for modules even though they don't differ across architectures as the module payload would still need to be tracked by a digest.

**Comment by nitrocode** [comment]
@cam72cam Oh then maybe we're talking about 2 different things.

The lock file seems to be a good fit for downloading and ensuring the SHASUM of the binaries are correct for the architecture given.

The SHA versioning is good for ensuring the version of the module is downloaded correctly IMHO.

---

If you folks believe that we can solve the problems of the lockfiles and SHA versioning of modules, then I'm all for lockfiles and committing them. 

For now, if https://github.com/opentofu/opentofu/issues/586 is completed prior to the lockfiles getting fixed, then we'll still be at a loss here.

**Comment by nitrocode** [comment]
@skyzyx sure, if it's possible to fix the lockfiles and make it easier, then I'm all for it.

A CI task would be an extra step. I'd prefer if the lock worked like npms lock or yarns lock. Once you commit the file, no manual steps are needed to alter it. You push and it works locally and in the CI process. I imagine this would require locking for common architectures by default.

**Comment by stevehipwell** [comment]
> If you folks believe that we can solve the problems of the lockfiles and SHA versioning of modules using lockfiles, then I'm all for lockfiles and committing them.

What problem? The UX could be improved for local changes, but lock files managed by tools like Dependabot are kept correctly configured for all present architectures.

@skyzyx I experimented with pre-commit hooks but the lock file generation is so slow that it's not really an option, so we just make sure we fail fast.

**Comment by nitrocode** [comment]
Ah this may be debating semantics now.

The root issue of the UX problem with lockfiles which prevents many users from committing lockfiles.

**Comment by cam72cam** [comment]
Spoke with the core team today.  Our plan is to patch the registry to lock tags to their initial hash and potentially provide a mechanism to update that (by maintainers).

**Comment by nitrocode** [comment]
Thanks for considering.

Do you mean they can lock tags to the shasum of the commit but then they will also have the ability to update the same tag to a different commit ?

**Comment by cam72cam** [comment]
Correct.  By default it will be idempotent and pointing directly at the hash we originally detect.  We will provide a path to refreshing the commit hashes for a given tag/repository.

Instead of returning https://github.com/owner/repo?ref=tag, the registry will return https://github.com/owner/repo?ref=commit-hash

**Comment by nitrocode** [comment]
But don't we want to lock the tag to a commit and prevent it from changing? 

Sorry for the silly question, but how does this change help with pinning modules to a sha ? Does this mean that we can set the module version to a sha once the registry change is made? 

Thank you

**Comment by stevehipwell** [comment]
@nitrocode I think they're saying that if you're accessing modules via the registry (I'm not sure if this would/could also work for direct git access by tag) the reference digest will be computed and the digest will be fixed even if the tag is moved. If so this would function much like how Go modules work out of the box.

**Comment by nitrocode** [comment]
Good to know, thanks for clarifying! 

I saw this today too which pins module sources as hashes in the terraform lockfile

https://github.com/ned1313/terrahash

**Comment by cam72cam** [comment]
Closed by https://github.com/opentofu/registry/pull/4113
